spicy/toolchain/src/compiler/validator.cc

```diff
@@ -198,6 +194,58 @@ struct VisitorPre : visitor::PreOrder, hilti::validator::VisitorMixIn {
 };
 
 struct VisitorPost : visitor::PreOrder, hilti::validator::VisitorMixIn {
+    /**
+     * Ensures that the node represented by tag can have all of the provided
+     * attributes. This does not use any context, if more information is needed,
+     * then do the check elsewhere.
+     */
+    void checkNodeAttributes(node::Tag tag, AttributeSet* attributes, const std::string_view& where) {
+        auto it = allowed_attrs.find(tag);
+
+        if ( ! attributes )
+            return;
+
+        if ( it == allowed_attrs.end() ) {
+            if ( ! attributes->attributes().empty() )
+                error(hilti::util::fmt("No attributes expected in %s", where), attributes);
+
+            return;
+        }
+
+        auto allowed = it->second;
+
+        for ( const auto& attr : attributes->attributes() )
+            if ( allowed.find(attr->kind()) == allowed.end() )
+                error(hilti::util::fmt("invalid attribute '%s' in %s", attr->attributeName(), where), attr);
+    }
+
+    /**
+     * Ensures that the type represented by typeTag can be within a field with
+     * the provided attributes. This is necessary since most attributes will apply
+     * to the field but not its type, so this gives a bit more context-sensitive
+     * validation for a common case.
+     */
+    void validateFieldTypeAttributes(node::Tag typeTag, AttributeSet* attributes, const std::string_view& clazz) {
+        auto it = field_type_allowed_attrs.find(typeTag);
+
+        if ( ! attributes )
+            return;
+
+        std::unordered_set<hilti::Attribute::Kind> allowed =
```

Comment:
Similar as above, would move the able out of there to the top of `validator.cc`.